sdk/core/azure-core-amqp/samples/internal/local_client_async_sample/local_client_async_sample.cpp

```diff
@@ -31,28 +31,30 @@ int main()
       session.CreateMessageSender("localhost/ingress", senderOptions, nullptr)};
 
   // Open the connection to the remote.
-  sender.Open();
+  if (sender.Open())
+  {
 
-  auto timeStart = std::chrono::high_resolution_clock::now();
+    auto timeStart = std::chrono::high_resolution_clock::now();
 
-  constexpr int maxMessageSendCount = 1000;
+    constexpr int maxMessageSendCount = 1000;
 
-  Azure::Core::Amqp::Models::AmqpMessage message;
-  message.SetBody(Azure::Core::Amqp::Models::AmqpBinaryData{'H', 'e', 'l', 'l', 'o'});
+    Azure::Core::Amqp::Models::AmqpMessage message;
+    message.SetBody(Azure::Core::Amqp::Models::AmqpBinaryData{'H', 'e', 'l', 'l', 'o'});
 
-  int messageSendCount = 0;
-  while (messageSendCount < maxMessageSendCount)
-  {
-    auto result = sender.Send(message);
-    messageSendCount += 1;
-  }
+    int messageSendCount = 0;
+    while (messageSendCount < maxMessageSendCount)
+    {
+      auto result = sender.Send(message);
+      messageSendCount += 1;
+    }
 
-  auto timeEnd = std::chrono::high_resolution_clock::now();
-  std::chrono::nanoseconds timeDiff = timeEnd - timeStart;
+    auto timeEnd = std::chrono::high_resolution_clock::now();
+    std::chrono::nanoseconds timeDiff = timeEnd - timeStart;
 
-  std::cout << "Sent " << messageSendCount << " in "
-            << std::chrono::duration_cast<std::chrono::milliseconds>(timeDiff).count()
-            << " milliseconds" << std::endl;
+    std::cout << "Sent " << messageSendCount << " in "
+              << std::chrono::duration_cast<std::chrono::milliseconds>(timeDiff).count()
+              << " milliseconds" << std::endl;
 
-  sender.Close();
+    sender.Close();
```

Comment:
If an error happens the destructor runs, does that ensure Close(). Its not clear who owns calling Close()

Reply:
If you called `Open`, then you are responsible for calling `Close` before the sender goes out of scope.

If you fail to call `Close`, then the destructor will call `abort()` and terminate the process. 

Unfortunately, the destructor *cannot* clean up opened resources in the destructor because C++ compilers assume that at the time a destructor is called that there are no references to the current object. The problem is that there sometimes ARE references to the current object (for instance, the sender has provided callbacks to the underlying uAMQP message sender object).

It might be possible to handle this using std::weak_reference and an intermediate object, but mandating a `Close` method is a dramatically simpler solution.